src/rust/engine/graph/src/lib.rs

```diff
@@ -137,7 +143,28 @@ impl<N: Node> InnerGraph<N> {
     // Search for an existing path from dst to src.
     let mut roots = VecDeque::new();
     roots.push_back(root);
-    self.walk(roots, direction).any(|eid| eid == needle)
+    let mut walk = self.walk(roots, direction);
+    let contains_cycle = walk.any(|eid| eid == needle);
+    if !contains_cycle {
+      return CycleType::None;
+    }
+    let dirty_nodes: HashSet<_> = walk
```

Comment:
This will over-approximate whether the path that forms the cycle _involved_ any dirty nodes, which might cause a bunch more stuff to get cleared than needs to be. An alternative would be to switch `walk` to something that actually records paths, or to re-run the walk in a way that will records paths once we have definitely detected a cycle?

When I started looking at this problem, I dove into trying to do it offline, and added the following method that might be useful here to find the cyclic path:
```
///
/// Compute and return one shortest path from `src` to any of `dsts`.
///
fn shortest_path(
  graph: &DiGraph<(), f32, u32>,
  src: &EntryId,
  dsts: &[EntryId],
) -> Result<Vec<EntryId>, String> {
  // NB: We use Bellman Ford because it actually records paths, unlike Dijkstra's.
  let (path_weights, paths) = petgraph::algo::bellman_ford(graph, *src).unwrap_or_else(|e| {
    panic!(
      "There should not be any negative edge weights. Got: {:?}",
      e
    )
  });

  // Find the root with the shortest path weight.
  let minimum_path_id = dsts
    .iter()
    .min_by_key(|&dst| path_weights[dst.index()] as usize)
    .ok_or_else(|| format!("None of {:?} were reachable from {:?}.", dsts, src))?;

  // Collect the path by walking through the `paths` Vec, which contains the indexes of
  // predecessor Nodes along a path to the src.
  let mut next_id = *minimum_path_id;
  let mut path = Vec::new();
  path.push(next_id);
  while let Some(current_id) = paths[next_id.index()] {
    path.push(current_id);
    if current_id == *src {
      break;
    }
    next_id = current_id;
  }
  Ok(path)
}
```

Reply:
(I guess that this isn't a blocker, but it would be good to follow up on.)